h2o-core/src/main/java/hex/leaderboard/Leaderboard.java

```diff
@@ -606,57 +612,43 @@ private static Model[] getModelsFromKeys(Key<Model>[] modelKeys) {
     return models;
   }
 
-  /**
-   * @return list of keys of models sorted by the given metric, fetched from the DKV
-   */
-  private Key<Model>[] sortModels(String metric) {
-    Key<Model>[] models = getModelKeys();
-    boolean decreasing = !isLossFunction(metric);
-    List<Key<Model>> newModelsSorted = ModelMetrics.sortModelsByMetric(metric, decreasing, Arrays.asList(models));
-    return newModelsSorted.toArray(new Key[0]);
-  }
-
   /**
    * Sort by metric on the leaderboard/test set or default model metrics.
    */
-  private Key<Model>[] sortModelKeys(Key<Model>[] modelKeys) {
-    final List<Key<Model>> sortedModelKeys;
-    boolean sortDecreasing = !isLossFunction(_sort_metric);
-    final Frame leaderboardFrame = leaderboardFrame();
-    try {
-      if (leaderboardFrame == null) {
-        sortedModelKeys = ModelMetrics.sortModelsByMetric(_sort_metric, sortDecreasing, Arrays.asList(modelKeys));
-      } else {
-        sortedModelKeys = ModelMetrics.sortModelsByMetric(leaderboardFrame, _sort_metric, sortDecreasing, Arrays.asList(modelKeys));
-      }
-    } catch (H2OIllegalArgumentException e) {
-      log.warn("ModelMetrics.sortModelsByMetric failed: " + e);
-      throw e;
+  private Key<Model>[] sortModelKeys(Key<Model>[] modelKeys, String sortMetric) {
+    boolean sortDecreasing = !isLossFunction(sortMetric);
+    Comparator<Key<Model>> cmp = Comparator.comparingDouble(mk -> getMetric(sortMetric, mk.get()));
+    if (sortDecreasing)
+      cmp = cmp.reversed();
+    modelKeys = modelKeys.clone();
+    Arrays.sort(modelKeys, cmp);
+    return modelKeys;
+  }
```

Comment:
This part makes sure we use the same data to sort as we show in the leaderboard.
` ModelMetrics.sortModelsByMetric(_sort_metric, sortDecreasing, Arrays.asList(modelKeys));` ignored the `scoring_data`.

Reply:
what do you mean by `sorting_data` here?
I'd like to understand the change, and the fact that you can't rely on `MetricsComparator` makes me wonder if there isn't anything wrong in the latter

Reply:
is it to ensure we alsways use `_leaderboard_model_metrics` ?

Reply:
ok, right! I guess that's because in `Leaderboard`, this fallback logic is done in `Leaderboard.getModelMetrics`

Reply:
Yes, you understand it correctly.

The `scoring_data` is a parameter in the `h2o.make_leaderboard` it can have one of the following values `AUTO | xval | valid | train` . Currently, `scoring_data` changes the metrics that we show in the leaderboard but, unfortunately, `ModelMetrics.sortModelsByMetric` has the `cv_metrics || valid_metrics || training_metrics` logic so the leaderboard might be sorted by data that are not shown (which is very confusing).

So this fixes that behavior - the leaderboard is now sorted by the metrics that are shown.